797 - Generate and add `README.md` to a zip file

nozomione opened the pull request:
## Issue Number
#797

## Purpose/Implementation Notes
This is the 2nd item in the stacked PR list for the linked issue.

The feature branch `feature/consolidate-readmes` was merged into this feature branch in #831.


The following changes are included:
- Added a new method `ComputedFile::get_portal_metadata_file`
- Adjusted `readme_file.get_file_contents` to use the new generic readme template 
- Added the following properties in `common`
- Updated the `create_portal_metadata` management command
- Added assertions to `test_create_portal_metadata` for the generated zip and `README.md` file 

## Types of changes

- New feature (non-breaking change which adds functionality)

## Functional tests
N/A

## Checklist
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots
N/A

davidsmejia reviewed (commented):
We should update this to create the computed file. Also the tests should check for the computed file.

nozomione commented on api/scpca_portal/readme_file.py:
```diff
@@ -38,3 +38,20 @@ def get_file_contents(download_config: Dict, project) -> str:
     ).strip()
+
+
+# NOTE: Temporarily added this until all consolidate readme PRs are finalized and merged
```
As per our 1:1 talk, I updated `NOTE` to `TODO` in the comments and instead of creating a new issue, I linked the PR that needed to be merged first. Thank you, David!

nozomione commented on api/scpca_portal/models/computed_file.py:
```diff
@@ -93,6 +93,22 @@ def get_local_sample_metadata_path(sample, download_config: Dict) -> Path:
+                    projects, common.GENERATED_PROJECT_DOWNLOAD_CONFIGURATIONS[-1]
+                ),
+            )
+
```
I've also updated the suffix `CONFIGURATIONS` to `CONFIG` of the following download configs and adjust the codebase accordingly:
 
- `GENERATED_PROJECT_DOWNLOAD_CONFIGURATIONS` 
- `GENERATED_SAMPLE_DOWNLOAD_CONFIG`

See the commit at d5ce75807aa2e0920de65f375fd8fa9dfb26dec1.

nozomione replied:
See the commit at 7b416aa60ca3ce55f5e7aa0b3ba6f243bd5f1c0c.

nozomione commented on api/scpca_portal/models/computed_file.py:
```diff
@@ -93,6 +93,22 @@ def get_local_sample_metadata_path(sample, download_config: Dict) -> Path:
+            zip_file.writestr(
+                readme_file.OUTPUT_NAME,
+                readme_file.get_portal_metadata_file_content(
+                    projects, common.GENERATED_PROJECT_DOWNLOAD_CONFIGURATIONS[-1]
```
I've also updated the suffix `CONFIGURATIONS` to `CONFIG` of the following download configs and adjust the codebase accordingly:
 
- `GENERATED_PROJECT_DOWNLOAD_CONFIGURATIONS` 
- `GENERATED_SAMPLE_DOWNLOAD_CONFIG`

See the commit at d5ce75807aa2e0920de65f375fd8fa9dfb26dec1.

nozomione commented:
I've removed `TODO` comments and adjusted the codeabse based on 1:1 (see at 3435b1833dc93fd321702e020229cf7aa3a78494 and 9a2728c221c63bafaa9b5d6562bc09bc85c42529), as well as updated the "Purpose/Implementation Notes" section of this PR.

This is ready for another look, thank you David!

nozomione commented on api/scpca_portal/common.py:
```diff
@@ -101,7 +101,7 @@
 LIBRARY_ID_KEY = "scpca_library_id"
 
-GENERATED_PROJECT_DOWNLOAD_CONFIGURATIONS = [
+GENERATED_PROJECT_DOWNLOAD_CONFIG = [
```
See at 1f5e0747bf1c3eb6e0b722d96b6361a3eead30f8.

nozomione commented on api/scpca_portal/readme_file.py:
```diff
@@ -10,10 +11,14 @@
 
 
-def get_file_contents(download_config: Dict, project) -> str:
+def get_file_contents(download_config: Dict, queryset: QuerySet) -> str:
```
Renamed `project` to `projects` and removed its type hint at ed9e0e415fc74cb97301e0ab357dc879b748869b.

nozomione commented on api/scpca_portal/readme_file.py:
```diff
@@ -10,10 +11,14 @@
-def get_file_contents(download_config: Dict, project) -> str:
+def get_file_contents(download_config: Dict, queryset: QuerySet) -> str:
     """Return newly generated readme file as a string for immediate writing to a zip archive."""
+    portal_metadata_only = download_config.get("portal_metadata_only", False)
```
I've removed this variable `portal_metadata_only` at ed9e0e415fc74cb97301e0ab357dc879b748869b.

nozomione commented on api/scpca_portal/readme_file.py:
```diff
@@ -10,10 +11,14 @@
     readme_template_key_parts = [download_config["modality"], download_config["format"]]
-    if download_config in common.GENERATED_PROJECT_DOWNLOAD_CONFIGURATIONS:
+
+    if portal_metadata_only:
```
I've switched `GENERATED_PORTAL_METADATA_DOWNLOAD_CONFIG` back from a dict to a list to match the project config check pattern at ed9e0e415fc74cb97301e0ab357dc879b748869b.

nozomione replied:
See at 5d81c0e64662b30217a0d2553fa7ce5993afd774.

nozomione commented on api/scpca_portal/readme_file.py:
```diff
@@ -30,6 +35,6 @@ def get_file_contents(download_config: Dict, project) -> str:
             "download_config": download_config,
             "contents_template": contents_template,
-            "projects": [project],
+            "projects": queryset if portal_metadata_only else [queryset],
```
I've used `QuerySet` to determine whether 'projects' should be passed as `[projects]` when dealing with a single project (no longer checking based on the `portal-metadata-only` key) at ed9e0e415fc74cb97301e0ab357dc879b748869b.

nozomione commented on api/scpca_portal/readme_file.py:
```diff
@@ -18,23 +19,43 @@
+        if is_portal_metadata
+        else common.GENERATED_PROJECT_DOWNLOAD_CONFIG
+    )
+
```
This code was already refactored and removed in a previous commit, so there's no change here, thank you!

davidsmejia commented on api/scpca_portal/models/library.py:
```diff
@@ -239,7 +239,7 @@ def get_zip_file_path(file_path: Path, download_config: Dict) -> Path:
 
         # Project output paths are relative to project directory
-        if download_config in common.GENERATED_PROJECT_DOWNLOAD_CONFIGURATIONS:
+        if download_config in common.GENERATED_PROJECT_DOWNLOAD_CONFIG:
```
```suggestion
        if download_config in common.GENERATED_PROJECT_DOWNLOAD_CONFIGS:
```

davidsmejia commented on api/scpca_portal/models/library.py:
```diff
@@ -121,7 +121,7 @@ def get_formats_from_file_paths(cls, file_paths: List[Path]) -> List[str]:
         cls, project, download_configuration: Dict
     ):  # -> QuerySet[Self]:
-        if download_configuration not in common.GENERATED_PROJECT_DOWNLOAD_CONFIGURATIONS:
+        if download_configuration not in common.GENERATED_PROJECT_DOWNLOAD_CONFIG:
```
```suggestion
        if download_configuration not in common.GENERATED_PROJECT_DOWNLOAD_CONFIGS:
```

nozomione replied:
See at 964fa78cae86bba9549ae971dcba8519eff843b2.

davidsmejia commented on api/scpca_portal/readme_file.py:
```diff
@@ -10,10 +11,13 @@
     readme_template_key_parts = [download_config["modality"], download_config["format"]]
-    if download_config in common.GENERATED_PROJECT_DOWNLOAD_CONFIGURATIONS:
+
+    if download_config in [common.GENERATED_PORTAL_METADATA_DOWNLOAD_CONFIG]:
```
```suggestion
    if download_config is common.GENERATED_PORTAL_METADATA_DOWNLOAD_CONFIG:
```

nozomione replied:
See at 964fa78cae86bba9549ae971dcba8519eff843b2.

davidsmejia reviewed (commented):
almost just need to make project configs also plural then we are good to go.

nozomione commented:
I've applied your input, and this PR is ready for another look. Thank you David!

davidsmejia commented on api/scpca_portal/readme_file.py:
```diff
@@ -10,10 +11,13 @@
 
 
-def get_file_contents(download_config: Dict, project) -> str:
+def get_file_contents(download_config: Dict, projects) -> str:
```
```suggestion
def get_file_contents(download_config: Dict, projects: Iterable) -> str:
```

youll need to import Iterable

nozomione replied:
See at 68053c986dc5b971a57bbee8039042a741dc20c8.

davidsmejia commented on api/scpca_portal/readme_file.py:
```diff
@@ -30,6 +34,6 @@ def get_file_contents(download_config: Dict, project) -> str:
             "download_config": download_config,
             "contents_template": contents_template,
-            "projects": [project],
+            "projects": projects if isinstance(projects, QuerySet) else [projects],
```
```suggestion
            "projects": projects,
```

You'll need to update other callers of this method so that it matches the iterable type.

nozomione replied:
See at 68053c986dc5b971a57bbee8039042a741dc20c8.

nozomione commented:
I've applied your feedback and this PR is ready for your review. Thank you David!

davidsmejia reviewed (approved):
LGTM

Let's create an issue now to update how we test for readme correctness here after `feature/consolidate-readmes` is merged in.

nozomione commented:
> Let's create an issue now to update how we test for readme correctness here after `feature/consolidate-readmes` is merged in.

I've filed a new issue here https://github.com/AlexsLemonade/scpca-portal/issues/869, thank you David!👍
